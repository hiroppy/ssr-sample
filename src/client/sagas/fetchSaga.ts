import { put, call, takeLatest, select } from 'redux-saga/effects';
import fetch from 'cross-fetch';
import {
  FETCH_SAGA_CODE,
  fetchSagaCode as FetchSagaCode,
  fetchSagaCodeSuccess,
  fetchSagaCodeFailure,
  ADD_LIKE,
  addLike as AddLike,
  addLikeSuccess,
  addLikeFailure
} from '../actions/fetchSaga';
import { State } from '../reducers/sagaPage';
import { getBaseUrl } from '../selectors';

function* fetchMethod({
  path,
  params,
  options
}: {
  path: string;
  params?: Record<string, any>;
  options?: Record<string, any>;
}) {
  const baseUrl = yield select(getBaseUrl);
  const url = new URL(`${baseUrl || window.origin}${path}`);

  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      url.searchParams.append(key, value || '');
    });
  }

  const res: Response = yield call(fetch, url.toString(), options);
  const { body }: { body: State['samples'] } = yield res.json();

  return body;
}

function* fetchSagaCode(action: ReturnType<typeof FetchSagaCode>) {
  try {
    const params = { maxLength: action.payload.maxLength };
    const body = yield fetchMethod({
      path: '/api/saga',
      params
    });

    yield put(fetchSagaCodeSuccess(body));
  } catch (err) {
    yield put(fetchSagaCodeFailure(err));
  }
}

function* addLike(action: ReturnType<typeof AddLike>) {
  const { id } = action.payload;
  const options = { method: 'POST' };

  try {
    // actually, this method doesn't run on Node.js so it's unnecessary to create a full url
    const body = yield fetchMethod({
      path: `/api/saga/${id}`,
      options
    });

    // need to optimize this because it's better to avoid fetching all data
    // currently, apollo's mutation uses `refetchQueries` so make the behavior the same.
    yield put(FetchSagaCode(null));
    yield put(addLikeSuccess(body));
  } catch (err) {
    yield put(addLikeFailure(err));
  }
}

export function* fetchSagaProcess() {
  yield takeLatest(FETCH_SAGA_CODE, fetchSagaCode);
  yield takeLatest(ADD_LIKE, addLike);
}
