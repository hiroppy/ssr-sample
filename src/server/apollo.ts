import * as express from 'express';
import { ApolloServer } from 'apollo-server-express';
import { getComplexity, simpleEstimator, fieldExtensionsEstimator } from 'graphql-query-complexity';
import { separateOperations } from 'graphql';
import { resolvers, schema } from '../graphql/schema';
import { endpoint, limitCost } from '../graphql/constants';

export function apollo(app: express.Application) {
  const apollo = new ApolloServer({
    debug: process.env.NODE_ENV !== 'production',
    schema,
    resolvers,
    playground:
      process.env.NODE_ENV !== 'production'
        ? {
            endpoint
          }
        : false,
    plugins: [
      {
        requestDidStart: () => ({
          didResolveOperation({ request, document }) {
            const complexity = getComplexity({
              schema,
              query: request.operationName
                ? separateOperations(document)[request.operationName]
                : document,
              variables: request.variables,
              estimators: [fieldExtensionsEstimator(), simpleEstimator({ defaultComplexity: 1 })]
            });

            if (complexity >= limitCost) {
              throw new Error(`${complexity} is over ${limitCost}`);
            }
            console.log('Used query complexity points:', complexity);
          },
          didEncounterErrors({ errors, request }) {
            // e.g. add a logger
            console.error(errors);
            console.error(request);
          }
        })
      }
    ]
  });

  apollo.applyMiddleware({ app, path: endpoint });
}
