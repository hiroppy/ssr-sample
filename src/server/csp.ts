import { Request, Response, NextFunction } from 'express';
import * as helmet from 'helmet';
import nanoid from 'nanoid';
import ua from 'ua-parser-js';

export function generateNonceId(req: Request, res: Response, next: NextFunction) {
  res.locals.nonce = Buffer.from(nanoid(32)).toString('base64');
  next();
}

// google font: https://stackoverflow.com/a/34576000/7014700
const baseDirectives: helmet.IHelmetContentSecurityPolicyDirectives = {
  defaultSrc: ["'self'"],
  styleSrc: ["'unsafe-inline'", 'fonts.googleapis.com'], // for styled-components
  fontSrc: ["'self'", 'data: fonts.gstatic.com'],
  imgSrc: ["'self'", 'img.shields.io'], // for README
  connectSrc: ["'self'", 'img.shields.io', 'fonts.googleapis.com', 'fonts.gstatic.com'], // for service-worker
  workerSrc: ["'self'"]
};

// chrome, firefox
const lv3Directives: helmet.IHelmetContentSecurityPolicyDirectives = {
  ...baseDirectives,
  scriptSrc: [(req, res) => `'nonce-${res.locals.nonce}'`, "'strict-dynamic'", "'unsafe-eval'"]
};

// safari
// TODO: cannot load a file to be imported as dynamic because cannot use strict-dynamic
const lv2Directives: helmet.IHelmetContentSecurityPolicyDirectives = {
  ...baseDirectives,
  scriptSrc: [
    "'self",
    (req, res) => `'nonce-${res.locals.nonce}'`,
    "'unsafe-eval'",
    "'unsafe-inline'"
  ]
};

export function csp(req: Request, res: Response, next: NextFunction) {
  const directives = ['Chrome', 'Firefox'].includes(
    // TODO: fix
    (ua as any)(req.headers['user-agent']).browser.name
  )
    ? lv3Directives
    : lv2Directives;

  helmet.contentSecurityPolicy({ directives })(req, res, next);
}
