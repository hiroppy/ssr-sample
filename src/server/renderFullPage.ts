type Params = {
  meta: string;
  body: string;
  style: string;
  preloadedState: string;
  graphql: string;
  nonce: string;
  scripts: string;
};

const escape = (str: string) => {
  return str
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
};

export const renderFullPage = ({
  meta,
  body,
  style,
  scripts,
  preloadedState,
  graphql,
  nonce
}: Params) => {
  return `<!DOCTYPE html>
    <html lang="en">
      <head>
        <meta charSet="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="Description" content="introducing SPA and SSR">
        <meta property="csp-nonce" content="${nonce}">
        ${meta}
        <link rel="manifest" href="/manifest.webmanifest">
        <link href="https://fonts.googleapis.com/css?family=Muli&display=swap" rel="stylesheet"></link>
        ${style}
      </head>
      <body>
        ${body}
        <script nonce="${nonce}" id="initial-data" type="text/plain" data-json="${escape(
    preloadedState
  )}"></script>
        <script nonce="${nonce}">window.__APOLLO_STATE__=${graphql}</script>
        ${scripts}
      </body>
    </html>
  `.trim();
};
