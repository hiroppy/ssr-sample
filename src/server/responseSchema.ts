import raw from 'raw.macro';

export type Samples = typeof apolloSamples;

// export type Themes = {
//   [key in 'dark' | 'light']: {
//     main: string;
//     sub: string;
//   };
// };

// TODO: fix
// need to create babel-plugin

export const sagaSamples = [
  {
    id: 1,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 2,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 3,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 4,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  }
];

export const apolloSamples = [
  {
    id: 1,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 2,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 3,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 4,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  }
];
