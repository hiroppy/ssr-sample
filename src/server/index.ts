import { join } from 'path';
import cluster from 'cluster';
import { cpus } from 'os';
import { config } from 'dotenv';
import { runServer } from './server';

const isProd = process.env.NODE_ENV === 'production';

// If you compile server code with webpack, this is unnecessary.
config({
  path: isProd ? join(__dirname, '../../../.env.prod') : join(__dirname, '../../.env.dev')
});

if (isProd) {
  const numCPUs = cpus().length;

  if (cluster.isMaster) {
    [...new Array(numCPUs)].forEach(() => cluster.fork());

    // cluster manager
    cluster.on('exit', (worker, code, signal) => {
      console.log(`Restarting ${worker.process.pid}. ${code || signal}`);
      cluster.fork();
    });
  } else {
    runServer();
  }
} else {
  runServer();
}

process.on('uncaughtException', (err) => {
  console.error(err);
  process.exit(1);
});

process.on('unhandledRejection', (err) => {
  console.error(err);
});
