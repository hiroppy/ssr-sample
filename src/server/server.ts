/* eslint @typescript-eslint/no-var-requires: 0 */

import { createServer } from 'http';
import express from 'express';
import * as bodyParser from 'body-parser';
import compression from 'compression';
import { generateNonceId, csp } from './csp';
import { apollo } from './apollo';
import { router } from './router';

export function runServer() {
  const app = express();
  const port = process.env.PORT || 3000;

  app.use(bodyParser.urlencoded({ extended: true }));
  app.use(bodyParser.json());

  // CSP
  app.use(generateNonceId);
  app.use(csp);

  // compression
  app.use(compression({ level: 5 }));

  // HMR
  if (process.env.NODE_ENV === 'development') {
    const webpack = require('webpack');
    const webpackHotMiddleware = require('webpack-hot-middleware');
    const webpackDevMiddleware = require('webpack-dev-middleware');
    const config = require('../../webpack.client.config');
    const compiler = webpack(config);

    app.use(webpackHotMiddleware(compiler));
    app.use(
      webpackDevMiddleware(compiler, {
        publicPath: config.output.publicPath,
        writeToDisk(filePath: string) {
          return /loadable-stats/.test(filePath);
        }
      })
    );
  }

  // setup apollo
  apollo(app);

  // register routes
  router(app);

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'test') {
    const server = createServer(app);

    server.listen(port, () => {
      console.log(`Listening on ${port}`);
    });

    // https://stackoverflow.com/a/48710483/7014700
    server.on('error', (err: NodeJS.ErrnoException) => {
      if (err.syscall !== 'listen') throw err;

      const bind = typeof port === 'string' ? `Pipe ${port}` : `Port ${port}`;

      switch (err.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw err;
      }
    });
  }

  return app;
}
