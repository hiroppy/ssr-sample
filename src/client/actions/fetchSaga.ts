import { loadSagaPage } from './pages';
import { sagaSamples } from '../../server/responseSchema';

export const FETCH_SAGA_CODE = 'FETCH_SAGA_CODE';
export const fetchSagaCode = (maxLength: Parameters<typeof loadSagaPage>[0]) =>
  ({
    type: FETCH_SAGA_CODE,
    payload: {
      maxLength
    }
  } as const);

export const FETCH_SAGA_CODE_SUCCESS = 'FETCH_SAGA_CODE_SUCCESS';
export const fetchSagaCodeSuccess = (samples: typeof sagaSamples) =>
  ({
    type: FETCH_SAGA_CODE_SUCCESS,
    payload: {
      samples
    }
  } as const);

export const FETCH_SAGA_CODE_FAILURE = 'FETCH_SAGA_CODE_FAILURE';
export const fetchSagaCodeFailure = (err: Error) =>
  ({
    type: FETCH_SAGA_CODE_FAILURE,
    payload: {
      err
    }
  } as const);

export const ADD_LIKE = 'ADD_LIKE';
export const addLike = (id: number) =>
  ({
    type: ADD_LIKE,
    payload: {
      id
    }
  } as const);

export const ADD_LIKE_SUCCESS = 'ADD_LIKE_SUCCESS';
export const addLikeSuccess = (res: typeof sagaSamples[0]) =>
  ({
    type: ADD_LIKE_SUCCESS,
    payload: {
      res
    }
  } as const);

export const ADD_LIKE_FAILURE = 'ADD_LIKE_FAILURE';
export const addLikeFailure = (err: Error) =>
  ({
    type: ADD_LIKE_FAILURE,
    payload: {
      err
    }
  } as const);

export type Actions =
  | ReturnType<typeof fetchSagaCode>
  | ReturnType<typeof fetchSagaCodeSuccess>
  | ReturnType<typeof fetchSagaCodeFailure>
  | ReturnType<typeof addLike>
  | ReturnType<typeof addLikeSuccess>
  | ReturnType<typeof addLikeFailure>;
