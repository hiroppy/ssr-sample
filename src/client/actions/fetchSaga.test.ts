import * as actions from './fetchSaga';
import { sagaSamples } from '../../server/responseSchema';

test("should create an action to fetch saga's code", () => {
  expect(actions.fetchSagaCode('2')).toMatchSnapshot();
});

test("should create a success action to fetch saga's code", () => {
  expect(actions.fetchSagaCodeSuccess(sagaSamples)).toMatchSnapshot();
});

test("should create a failure action to fetch saga's code", () => {
  expect(actions.fetchSagaCodeFailure(new Error())).toMatchSnapshot();
});

test('should create an action to add like count', () => {
  expect(actions.addLike(1)).toMatchSnapshot();
});

test('should create a success action to add like count', () => {
  expect(actions.addLikeSuccess(sagaSamples[0])).toMatchSnapshot();
});

test('should create a failure action to add like count', () => {
  expect(actions.addLikeFailure(new Error())).toMatchSnapshot();
});
