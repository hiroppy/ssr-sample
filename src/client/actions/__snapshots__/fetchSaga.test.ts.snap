// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a failure action to add like count 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "ADD_LIKE_FAILURE",
}
`;

exports[`should create a failure action to fetch saga's code 1`] = `
Object {
  "payload": Object {
    "err": [Error],
  },
  "type": "FETCH_SAGA_CODE_FAILURE",
}
`;

exports[`should create a success action to add like count 1`] = `
Object {
  "payload": Object {
    "res": Object {
      "code": "test-code",
      "description": "aaaa",
      "id": 1,
      "likeCount": 0,
      "name": "pages.tsx",
    },
  },
  "type": "ADD_LIKE_SUCCESS",
}
`;

exports[`should create a success action to fetch saga's code 1`] = `
Object {
  "payload": Object {
    "samples": Array [
      Object {
        "code": "test-code",
        "description": "aaaa",
        "id": 1,
        "likeCount": 0,
        "name": "pages.tsx",
      },
      Object {
        "code": "test-code",
        "description": "aaaa",
        "id": 2,
        "likeCount": 0,
        "name": "pages.tsx",
      },
      Object {
        "code": "test-code",
        "description": "aaaa",
        "id": 3,
        "likeCount": 0,
        "name": "pages.tsx",
      },
      Object {
        "code": "test-code",
        "description": "aaaa",
        "id": 4,
        "likeCount": 0,
        "name": "pages.tsx",
      },
    ],
  },
  "type": "FETCH_SAGA_CODE_SUCCESS",
}
`;

exports[`should create an action to add like count 1`] = `
Object {
  "payload": Object {
    "id": 1,
  },
  "type": "ADD_LIKE",
}
`;

exports[`should create an action to fetch saga's code 1`] = `
Object {
  "payload": Object {
    "maxLength": "2",
  },
  "type": "FETCH_SAGA_CODE",
}
`;
